{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "severe-customer",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "attempted-supervisor",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import os\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "responsible-florida",
   "metadata": {},
   "source": [
    "## Def describe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "interesting-sacrifice",
   "metadata": {},
   "outputs": [],
   "source": [
    "def describe(df, random=False):\n",
    "    print(df.shape)\n",
    "    print(\"--- First 5 entries ---\")\n",
    "    display(df.head())\n",
    "    if random:\n",
    "        print('--- Random subset ---')\n",
    "        display(df.sample(5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "brilliant-jurisdiction",
   "metadata": {},
   "source": [
    "## Read one2one h5ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "thick-georgia",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/olga/miniconda3/envs/immune-evolution/lib/python3.7/site-packages/anndata/_core/anndata.py:1094: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead\n",
      "  if not is_categorical(df_full[k]):\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 89795 × 10560\n",
       "    obs: 'index', 'age', 'cell_barcode', 'cell_ontology_class', 'cell_ontology_id', 'channel', 'free_annotation', 'individual', 'sample', 'sequencing_run', 'sex', 'species', 'species_batch', 'species_latin', 'tissue', 'narrow_group', 'broad_group', 'compartment_group', 'compartment_narrow', 'channel_cleaned', 'batch', 'n_genes', 'n_counts', 'species_batch_v2', 'compartment_broad', 'compartment_broad_narrow', 'compartment_species', 'compartment_narrow_species', 'common_individual_id'\n",
       "    var: 'bat__gene_name', 'mouse_lemur__gene_name-bat', 'mouse__gene_name-bat', 'mouse_lemur__gene_name_x-hlm', 'mouse__gene_name_x-hlm', 'gene_ids-lemur-hlm', 'n_cells-mouse-hlm', 'mouse_lemur__gene_name_y-hlm', 'mouse__gene_name_y-hlm'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read(\n",
    "#     \"/home/phoenix/data_sm/code/immune-evolution/h5ads/human-lemur-mouse-bat__lung_only_copy.h5ad\"\n",
    "    \"/home/olga/data_sm/immune-evolution/h5ads/human-lemur-mouse-bat/human-lemur-mouse-bat__lung_only.h5ad\"\n",
    ")\n",
    "adata.obs = adata.obs.reset_index().set_index('cell_id')\n",
    "SHARED_CELLTYPES = ['Alveolar Epithelial Type 2',\n",
    " 'B cell',\n",
    " 'Capillary',\n",
    " 'Dendritic',\n",
    " 'Fibroblast',\n",
    " 'Macrophage',\n",
    " 'Monocyte',\n",
    " 'Natural Killer T cell',\n",
    " 'Smooth Muscle and Myofibroblast',\n",
    " 'T cell']\n",
    "len(SHARED_CELLTYPES)\n",
    "\n",
    "adata_shared = adata[adata.obs.broad_group.isin(SHARED_CELLTYPES)]\n",
    "adata_shared"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "terminal-bundle",
   "metadata": {},
   "source": [
    "# Read parquet files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "boolean-leadership",
   "metadata": {},
   "source": [
    "## File paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "green-coordinator",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/home/olga/data_sm/immune-evolution/kmer-signatures/2--test-human/2--single-cell-kmers/alphabet-dayhoff__ksize-51__scaled-10/hash2kmer__unique_kmers_per_celltype.parquet',\n",
       " '/home/olga/data_sm/immune-evolution/kmer-signatures/3--test-bat/2--single-cell-kmers/alphabet-dayhoff__ksize-51__scaled-10/hash2kmer__unique_kmers_per_celltype.parquet',\n",
       " '/home/olga/data_sm/immune-evolution/kmer-signatures/4--test-lemur/2--single-cell-kmers/alphabet-dayhoff__ksize-51__scaled-10/hash2kmer__unique_kmers_per_celltype.parquet',\n",
       " '/home/olga/data_sm/immune-evolution/kmer-signatures/1--train-mouse/2--single-cell-kmers/alphabet-dayhoff__ksize-51__scaled-10/hash2kmer__unique_kmers_per_celltype.parquet']"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sketch_id = 'alphabet-dayhoff__ksize-51__scaled-10'\n",
    "sig_outdir_base = \"/home/olga/data_sm/immune-evolution/kmer-signatures\"\n",
    "\n",
    "\n",
    "parquets = glob.glob(f'/home/olga/data_sm/immune-evolution/kmer-signatures/*/2--single-cell-kmers/{sketch_id}/hash2kmer__unique_kmers_per_celltype.parquet')\n",
    "parquets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "adult-resolution",
   "metadata": {},
   "outputs": [],
   "source": [
    "# parquets = glob.glob(f'/home/olga/data_sm/immune-evolution/kmer-signatures/*/2--single-cell-kmers/*/hash2kmer__unique_kmers_per_celltype.parquet')\n",
    "# parquets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "placed-acquisition",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 olga czb 2.4G May  3 11:03 /home/olga/data_sm/immune-evolution/kmer-signatures/2--test-human/2--single-cell-kmers/alphabet-dayhoff__ksize-51__scaled-10/hash2kmer__unique_kmers_per_celltype.parquet\n",
      "CPU times: user 1min 2s, sys: 40.4 s, total: 1min 43s\n",
      "Wall time: 2min 56s\n",
      "-rw-r--r-- 1 olga czb 5.0G May 10 13:26 /home/olga/data_sm/immune-evolution/kmer-signatures/3--test-bat/2--single-cell-kmers/alphabet-dayhoff__ksize-51__scaled-10/hash2kmer__unique_kmers_per_celltype.parquet\n",
      "CPU times: user 2min 53s, sys: 1min 3s, total: 3min 56s\n",
      "Wall time: 3min 52s\n",
      "-rw-r--r-- 1 olga czb 149M Apr 30 13:26 /home/olga/data_sm/immune-evolution/kmer-signatures/4--test-lemur/2--single-cell-kmers/alphabet-dayhoff__ksize-51__scaled-10/hash2kmer__unique_kmers_per_celltype.parquet\n",
      "CPU times: user 3.89 s, sys: 2.02 s, total: 5.9 s\n",
      "Wall time: 4.11 s\n",
      "-rw-r--r-- 1 olga czb 3.3G May  3 11:12 /home/olga/data_sm/immune-evolution/kmer-signatures/1--train-mouse/2--single-cell-kmers/alphabet-dayhoff__ksize-51__scaled-10/hash2kmer__unique_kmers_per_celltype.parquet\n",
      "CPU times: user 1min 25s, sys: 39.6 s, total: 2min 4s\n",
      "Wall time: 1min 27s\n",
      "(148123511, 13)\n",
      "--- First 5 entries ---\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer_in_sequence</th>\n",
       "      <th>kmer_in_alphabet</th>\n",
       "      <th>hashval</th>\n",
       "      <th>read_name</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>species</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>sketch_id</th>\n",
       "      <th>moltype</th>\n",
       "      <th>ksize</th>\n",
       "      <th>scaled</th>\n",
       "      <th>alignment_status</th>\n",
       "      <th>groundtruth_celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DLNVKPKTIKTLEENLG</td>\n",
       "      <td>cecedbdbedbeccceb</td>\n",
       "      <td>550348301278027135</td>\n",
       "      <td>A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IKTLEENLGITIQDIGV</td>\n",
       "      <td>edbecccebebeccebe</td>\n",
       "      <td>974148631326387115</td>\n",
       "      <td>A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CRKLKLDPFLTPYTKIN</td>\n",
       "      <td>addedecbfebbfbdec</td>\n",
       "      <td>1662986983738689810</td>\n",
       "      <td>A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RKLKLDPFLTPYTKINS</td>\n",
       "      <td>ddedecbfebbfbdecb</td>\n",
       "      <td>799720167227311772</td>\n",
       "      <td>A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>LDPFLTPYTKINSRWIK</td>\n",
       "      <td>ecbfebbfbdecbdfed</td>\n",
       "      <td>537198855570371996</td>\n",
       "      <td>A00111:155:HFCWYDMXX:2:1360:14886:30185\\tCB:Z:...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    kmer_in_sequence   kmer_in_alphabet              hashval  \\\n",
       "0  DLNVKPKTIKTLEENLG  cecedbdbedbeccceb   550348301278027135   \n",
       "1  IKTLEENLGITIQDIGV  edbecccebebeccebe   974148631326387115   \n",
       "3  CRKLKLDPFLTPYTKIN  addedecbfebbfbdec  1662986983738689810   \n",
       "4  RKLKLDPFLTPYTKINS  ddedecbfebbfbdecb   799720167227311772   \n",
       "7  LDPFLTPYTKINSRWIK  ecbfebbfbdecbdfed   537198855570371996   \n",
       "\n",
       "                                           read_name gene_name species  \\\n",
       "0  A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...      None   human   \n",
       "1  A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...      None   human   \n",
       "3  A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...      None   human   \n",
       "4  A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...      None   human   \n",
       "7  A00111:155:HFCWYDMXX:2:1360:14886:30185\\tCB:Z:...      None   human   \n",
       "\n",
       "                  cell_id                              sketch_id  moltype  \\\n",
       "0  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "1  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "3  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "4  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "7  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "\n",
       "  ksize  scaled alignment_status groundtruth_celltype  \n",
       "0    51      10        unaligned           Fibroblast  \n",
       "1    51      10        unaligned           Fibroblast  \n",
       "3    51      10        unaligned           Fibroblast  \n",
       "4    51      10        unaligned           Fibroblast  \n",
       "7    51      10        unaligned           Fibroblast  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 7min 13s, sys: 4min 15s, total: 11min 28s\n",
      "Wall time: 13min 42s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "dfs = []\n",
    "\n",
    "for parquet in parquets:\n",
    "    ! ls -lha $parquet\n",
    "    %time df = pd.read_parquet(parquet)\n",
    "    df = df.rename(columns={'broad_group': 'groundtruth_celltype'})\n",
    "    dfs.append(df)\n",
    "    \n",
    "hash2kmer = pd.concat(dfs)\n",
    "describe(hash2kmer)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "trying-surveillance",
   "metadata": {},
   "source": [
    "## Read predicted cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "occupied-affiliate",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(29750, 43)\n",
      "--- First 5 entries ---\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>similarity</th>\n",
       "      <th>mouse1_broad_group</th>\n",
       "      <th>search_method</th>\n",
       "      <th>database_type</th>\n",
       "      <th>sketch_id</th>\n",
       "      <th>mol_or_alpha</th>\n",
       "      <th>alphabet</th>\n",
       "      <th>ksize</th>\n",
       "      <th>sketch_style</th>\n",
       "      <th>sketch_value</th>\n",
       "      <th>...</th>\n",
       "      <th>bat_narrow_group</th>\n",
       "      <th>bat_broad_group</th>\n",
       "      <th>bat_compartment_group</th>\n",
       "      <th>human_channel</th>\n",
       "      <th>human_cell_ontology_class</th>\n",
       "      <th>human_narrow_group</th>\n",
       "      <th>human_broad_group</th>\n",
       "      <th>human_compartment_group</th>\n",
       "      <th>predicted_compartment</th>\n",
       "      <th>groundtruth_compartment</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>query_cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MACA_18m_F_LUNG_51__AAACCTGCAGCCTATA</th>\n",
       "      <td>0.119992</td>\n",
       "      <td>T cell</td>\n",
       "      <td>similarity_abundance</td>\n",
       "      <td>min-5-percent-shared-kmers-remove-common-kmers</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>alphabet</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>scaled</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>lymphoid</td>\n",
       "      <td>lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MACA_18m_F_LUNG_51__AAACCTGGTGTGGTTT</th>\n",
       "      <td>0.170558</td>\n",
       "      <td>B cell</td>\n",
       "      <td>similarity_abundance</td>\n",
       "      <td>min-5-percent-shared-kmers-remove-common-kmers</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>alphabet</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>scaled</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>lymphoid</td>\n",
       "      <td>lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MACA_18m_F_LUNG_51__AAACGGGAGTTTCCTT</th>\n",
       "      <td>0.080257</td>\n",
       "      <td>T cell</td>\n",
       "      <td>similarity_abundance</td>\n",
       "      <td>min-5-percent-shared-kmers-remove-common-kmers</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>alphabet</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>scaled</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>lymphoid</td>\n",
       "      <td>lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MACA_18m_F_LUNG_51__AAACGGGCAGTTCATG</th>\n",
       "      <td>0.263881</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>similarity_abundance</td>\n",
       "      <td>min-5-percent-shared-kmers-remove-common-kmers</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>alphabet</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>scaled</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>myeloid</td>\n",
       "      <td>myeloid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MACA_18m_F_LUNG_51__AAACGGGGTCATCCCT</th>\n",
       "      <td>0.326669</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>similarity_abundance</td>\n",
       "      <td>min-5-percent-shared-kmers-remove-common-kmers</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>alphabet</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>scaled</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>myeloid</td>\n",
       "      <td>myeloid</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      similarity mouse1_broad_group  \\\n",
       "query_cell_id                                                         \n",
       "MACA_18m_F_LUNG_51__AAACCTGCAGCCTATA    0.119992             T cell   \n",
       "MACA_18m_F_LUNG_51__AAACCTGGTGTGGTTT    0.170558             B cell   \n",
       "MACA_18m_F_LUNG_51__AAACGGGAGTTTCCTT    0.080257             T cell   \n",
       "MACA_18m_F_LUNG_51__AAACGGGCAGTTCATG    0.263881           Monocyte   \n",
       "MACA_18m_F_LUNG_51__AAACGGGGTCATCCCT    0.326669           Monocyte   \n",
       "\n",
       "                                             search_method  \\\n",
       "query_cell_id                                                \n",
       "MACA_18m_F_LUNG_51__AAACCTGCAGCCTATA  similarity_abundance   \n",
       "MACA_18m_F_LUNG_51__AAACCTGGTGTGGTTT  similarity_abundance   \n",
       "MACA_18m_F_LUNG_51__AAACGGGAGTTTCCTT  similarity_abundance   \n",
       "MACA_18m_F_LUNG_51__AAACGGGCAGTTCATG  similarity_abundance   \n",
       "MACA_18m_F_LUNG_51__AAACGGGGTCATCCCT  similarity_abundance   \n",
       "\n",
       "                                                                       database_type  \\\n",
       "query_cell_id                                                                          \n",
       "MACA_18m_F_LUNG_51__AAACCTGCAGCCTATA  min-5-percent-shared-kmers-remove-common-kmers   \n",
       "MACA_18m_F_LUNG_51__AAACCTGGTGTGGTTT  min-5-percent-shared-kmers-remove-common-kmers   \n",
       "MACA_18m_F_LUNG_51__AAACGGGAGTTTCCTT  min-5-percent-shared-kmers-remove-common-kmers   \n",
       "MACA_18m_F_LUNG_51__AAACGGGCAGTTCATG  min-5-percent-shared-kmers-remove-common-kmers   \n",
       "MACA_18m_F_LUNG_51__AAACGGGGTCATCCCT  min-5-percent-shared-kmers-remove-common-kmers   \n",
       "\n",
       "                                                                  sketch_id  \\\n",
       "query_cell_id                                                                 \n",
       "MACA_18m_F_LUNG_51__AAACCTGCAGCCTATA  alphabet-dayhoff__ksize-51__scaled-10   \n",
       "MACA_18m_F_LUNG_51__AAACCTGGTGTGGTTT  alphabet-dayhoff__ksize-51__scaled-10   \n",
       "MACA_18m_F_LUNG_51__AAACGGGAGTTTCCTT  alphabet-dayhoff__ksize-51__scaled-10   \n",
       "MACA_18m_F_LUNG_51__AAACGGGCAGTTCATG  alphabet-dayhoff__ksize-51__scaled-10   \n",
       "MACA_18m_F_LUNG_51__AAACGGGGTCATCCCT  alphabet-dayhoff__ksize-51__scaled-10   \n",
       "\n",
       "                                     mol_or_alpha alphabet  ksize  \\\n",
       "query_cell_id                                                       \n",
       "MACA_18m_F_LUNG_51__AAACCTGCAGCCTATA     alphabet  dayhoff     51   \n",
       "MACA_18m_F_LUNG_51__AAACCTGGTGTGGTTT     alphabet  dayhoff     51   \n",
       "MACA_18m_F_LUNG_51__AAACGGGAGTTTCCTT     alphabet  dayhoff     51   \n",
       "MACA_18m_F_LUNG_51__AAACGGGCAGTTCATG     alphabet  dayhoff     51   \n",
       "MACA_18m_F_LUNG_51__AAACGGGGTCATCCCT     alphabet  dayhoff     51   \n",
       "\n",
       "                                     sketch_style  sketch_value  ...  \\\n",
       "query_cell_id                                                    ...   \n",
       "MACA_18m_F_LUNG_51__AAACCTGCAGCCTATA       scaled            10  ...   \n",
       "MACA_18m_F_LUNG_51__AAACCTGGTGTGGTTT       scaled            10  ...   \n",
       "MACA_18m_F_LUNG_51__AAACGGGAGTTTCCTT       scaled            10  ...   \n",
       "MACA_18m_F_LUNG_51__AAACGGGCAGTTCATG       scaled            10  ...   \n",
       "MACA_18m_F_LUNG_51__AAACGGGGTCATCCCT       scaled            10  ...   \n",
       "\n",
       "                                     bat_narrow_group bat_broad_group  \\\n",
       "query_cell_id                                                           \n",
       "MACA_18m_F_LUNG_51__AAACCTGCAGCCTATA             None            None   \n",
       "MACA_18m_F_LUNG_51__AAACCTGGTGTGGTTT             None            None   \n",
       "MACA_18m_F_LUNG_51__AAACGGGAGTTTCCTT             None            None   \n",
       "MACA_18m_F_LUNG_51__AAACGGGCAGTTCATG             None            None   \n",
       "MACA_18m_F_LUNG_51__AAACGGGGTCATCCCT             None            None   \n",
       "\n",
       "                                     bat_compartment_group human_channel  \\\n",
       "query_cell_id                                                              \n",
       "MACA_18m_F_LUNG_51__AAACCTGCAGCCTATA                  None          None   \n",
       "MACA_18m_F_LUNG_51__AAACCTGGTGTGGTTT                  None          None   \n",
       "MACA_18m_F_LUNG_51__AAACGGGAGTTTCCTT                  None          None   \n",
       "MACA_18m_F_LUNG_51__AAACGGGCAGTTCATG                  None          None   \n",
       "MACA_18m_F_LUNG_51__AAACGGGGTCATCCCT                  None          None   \n",
       "\n",
       "                                     human_cell_ontology_class  \\\n",
       "query_cell_id                                                    \n",
       "MACA_18m_F_LUNG_51__AAACCTGCAGCCTATA                      None   \n",
       "MACA_18m_F_LUNG_51__AAACCTGGTGTGGTTT                      None   \n",
       "MACA_18m_F_LUNG_51__AAACGGGAGTTTCCTT                      None   \n",
       "MACA_18m_F_LUNG_51__AAACGGGCAGTTCATG                      None   \n",
       "MACA_18m_F_LUNG_51__AAACGGGGTCATCCCT                      None   \n",
       "\n",
       "                                     human_narrow_group human_broad_group  \\\n",
       "query_cell_id                                                               \n",
       "MACA_18m_F_LUNG_51__AAACCTGCAGCCTATA               None              None   \n",
       "MACA_18m_F_LUNG_51__AAACCTGGTGTGGTTT               None              None   \n",
       "MACA_18m_F_LUNG_51__AAACGGGAGTTTCCTT               None              None   \n",
       "MACA_18m_F_LUNG_51__AAACGGGCAGTTCATG               None              None   \n",
       "MACA_18m_F_LUNG_51__AAACGGGGTCATCCCT               None              None   \n",
       "\n",
       "                                     human_compartment_group  \\\n",
       "query_cell_id                                                  \n",
       "MACA_18m_F_LUNG_51__AAACCTGCAGCCTATA                    None   \n",
       "MACA_18m_F_LUNG_51__AAACCTGGTGTGGTTT                    None   \n",
       "MACA_18m_F_LUNG_51__AAACGGGAGTTTCCTT                    None   \n",
       "MACA_18m_F_LUNG_51__AAACGGGCAGTTCATG                    None   \n",
       "MACA_18m_F_LUNG_51__AAACGGGGTCATCCCT                    None   \n",
       "\n",
       "                                     predicted_compartment  \\\n",
       "query_cell_id                                                \n",
       "MACA_18m_F_LUNG_51__AAACCTGCAGCCTATA              lymphoid   \n",
       "MACA_18m_F_LUNG_51__AAACCTGGTGTGGTTT              lymphoid   \n",
       "MACA_18m_F_LUNG_51__AAACGGGAGTTTCCTT              lymphoid   \n",
       "MACA_18m_F_LUNG_51__AAACGGGCAGTTCATG               myeloid   \n",
       "MACA_18m_F_LUNG_51__AAACGGGGTCATCCCT               myeloid   \n",
       "\n",
       "                                     groundtruth_compartment  \n",
       "query_cell_id                                                 \n",
       "MACA_18m_F_LUNG_51__AAACCTGCAGCCTATA                lymphoid  \n",
       "MACA_18m_F_LUNG_51__AAACCTGGTGTGGTTT                lymphoid  \n",
       "MACA_18m_F_LUNG_51__AAACGGGAGTTTCCTT                lymphoid  \n",
       "MACA_18m_F_LUNG_51__AAACGGGCAGTTCATG                 myeloid  \n",
       "MACA_18m_F_LUNG_51__AAACGGGGTCATCCCT                 myeloid  \n",
       "\n",
       "[5 rows x 43 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "predicted_cells_min_5_percent = pd.read_parquet(\n",
    "    os.path.join(sig_outdir_base, \"aggregated-predicted-cells.parquet\")\n",
    ")\n",
    "predicted_cells_min_5_percent = predicted_cells_min_5_percent.query('sketch_id == @sketch_id')\n",
    "describe(predicted_cells_min_5_percent)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "personalized-friendship",
   "metadata": {},
   "source": [
    "### Join hash2kmer with predicted cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "ranging-merit",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(29750, 3)\n",
      "--- First 5 entries ---\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted_celltype</th>\n",
       "      <th>groundtruth_compartment</th>\n",
       "      <th>predicted_compartment</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>query_cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MACA_18M_F_LUNG_51__AAACCTGCAGCCTATA</th>\n",
       "      <td>T cell</td>\n",
       "      <td>lymphoid</td>\n",
       "      <td>lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MACA_18M_F_LUNG_51__AAACCTGGTGTGGTTT</th>\n",
       "      <td>B cell</td>\n",
       "      <td>lymphoid</td>\n",
       "      <td>lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MACA_18M_F_LUNG_51__AAACGGGAGTTTCCTT</th>\n",
       "      <td>T cell</td>\n",
       "      <td>lymphoid</td>\n",
       "      <td>lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MACA_18M_F_LUNG_51__AAACGGGCAGTTCATG</th>\n",
       "      <td>Monocyte</td>\n",
       "      <td>myeloid</td>\n",
       "      <td>myeloid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MACA_18M_F_LUNG_51__AAACGGGGTCATCCCT</th>\n",
       "      <td>Monocyte</td>\n",
       "      <td>myeloid</td>\n",
       "      <td>myeloid</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     predicted_celltype  \\\n",
       "query_cell_id                                             \n",
       "MACA_18M_F_LUNG_51__AAACCTGCAGCCTATA             T cell   \n",
       "MACA_18M_F_LUNG_51__AAACCTGGTGTGGTTT             B cell   \n",
       "MACA_18M_F_LUNG_51__AAACGGGAGTTTCCTT             T cell   \n",
       "MACA_18M_F_LUNG_51__AAACGGGCAGTTCATG           Monocyte   \n",
       "MACA_18M_F_LUNG_51__AAACGGGGTCATCCCT           Monocyte   \n",
       "\n",
       "                                     groundtruth_compartment  \\\n",
       "query_cell_id                                                  \n",
       "MACA_18M_F_LUNG_51__AAACCTGCAGCCTATA                lymphoid   \n",
       "MACA_18M_F_LUNG_51__AAACCTGGTGTGGTTT                lymphoid   \n",
       "MACA_18M_F_LUNG_51__AAACGGGAGTTTCCTT                lymphoid   \n",
       "MACA_18M_F_LUNG_51__AAACGGGCAGTTCATG                 myeloid   \n",
       "MACA_18M_F_LUNG_51__AAACGGGGTCATCCCT                 myeloid   \n",
       "\n",
       "                                     predicted_compartment  \n",
       "query_cell_id                                               \n",
       "MACA_18M_F_LUNG_51__AAACCTGCAGCCTATA              lymphoid  \n",
       "MACA_18M_F_LUNG_51__AAACCTGGTGTGGTTT              lymphoid  \n",
       "MACA_18M_F_LUNG_51__AAACGGGAGTTTCCTT              lymphoid  \n",
       "MACA_18M_F_LUNG_51__AAACGGGCAGTTCATG               myeloid  \n",
       "MACA_18M_F_LUNG_51__AAACGGGGTCATCCCT               myeloid  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "celltype_predictions = predicted_cells_min_5_percent[\n",
    "    [\"predicted_celltype\", \"groundtruth_compartment\", \"predicted_compartment\"]\n",
    "]\n",
    "celltype_predictions.index = celltype_predictions.index.str.upper()\n",
    "describe(celltype_predictions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "aware-surgeon",
   "metadata": {},
   "outputs": [],
   "source": [
    "hash2kmer.cell_id = hash2kmer.cell_id.str.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "facial-truth",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(148123511, 16)\n",
      "--- First 5 entries ---\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer_in_sequence</th>\n",
       "      <th>kmer_in_alphabet</th>\n",
       "      <th>hashval</th>\n",
       "      <th>read_name</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>species</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>sketch_id</th>\n",
       "      <th>moltype</th>\n",
       "      <th>ksize</th>\n",
       "      <th>scaled</th>\n",
       "      <th>alignment_status</th>\n",
       "      <th>groundtruth_celltype</th>\n",
       "      <th>predicted_celltype</th>\n",
       "      <th>groundtruth_compartment</th>\n",
       "      <th>predicted_compartment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DLNVKPKTIKTLEENLG</td>\n",
       "      <td>cecedbdbedbeccceb</td>\n",
       "      <td>550348301278027135</td>\n",
       "      <td>A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>stromal</td>\n",
       "      <td>stromal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IKTLEENLGITIQDIGV</td>\n",
       "      <td>edbecccebebeccebe</td>\n",
       "      <td>974148631326387115</td>\n",
       "      <td>A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>stromal</td>\n",
       "      <td>stromal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CRKLKLDPFLTPYTKIN</td>\n",
       "      <td>addedecbfebbfbdec</td>\n",
       "      <td>1662986983738689810</td>\n",
       "      <td>A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>stromal</td>\n",
       "      <td>stromal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RKLKLDPFLTPYTKINS</td>\n",
       "      <td>ddedecbfebbfbdecb</td>\n",
       "      <td>799720167227311772</td>\n",
       "      <td>A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>stromal</td>\n",
       "      <td>stromal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>LDPFLTPYTKINSRWIK</td>\n",
       "      <td>ecbfebbfbdecbdfed</td>\n",
       "      <td>537198855570371996</td>\n",
       "      <td>A00111:155:HFCWYDMXX:2:1360:14886:30185\\tCB:Z:...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>stromal</td>\n",
       "      <td>stromal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    kmer_in_sequence   kmer_in_alphabet              hashval  \\\n",
       "0  DLNVKPKTIKTLEENLG  cecedbdbedbeccceb   550348301278027135   \n",
       "1  IKTLEENLGITIQDIGV  edbecccebebeccebe   974148631326387115   \n",
       "3  CRKLKLDPFLTPYTKIN  addedecbfebbfbdec  1662986983738689810   \n",
       "4  RKLKLDPFLTPYTKINS  ddedecbfebbfbdecb   799720167227311772   \n",
       "7  LDPFLTPYTKINSRWIK  ecbfebbfbdecbdfed   537198855570371996   \n",
       "\n",
       "                                           read_name gene_name species  \\\n",
       "0  A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...      None   human   \n",
       "1  A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...      None   human   \n",
       "3  A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...      None   human   \n",
       "4  A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...      None   human   \n",
       "7  A00111:155:HFCWYDMXX:2:1360:14886:30185\\tCB:Z:...      None   human   \n",
       "\n",
       "                  cell_id                              sketch_id  moltype  \\\n",
       "0  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "1  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "3  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "4  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "7  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "\n",
       "  ksize  scaled alignment_status groundtruth_celltype predicted_celltype  \\\n",
       "0    51      10        unaligned           Fibroblast         Fibroblast   \n",
       "1    51      10        unaligned           Fibroblast         Fibroblast   \n",
       "3    51      10        unaligned           Fibroblast         Fibroblast   \n",
       "4    51      10        unaligned           Fibroblast         Fibroblast   \n",
       "7    51      10        unaligned           Fibroblast         Fibroblast   \n",
       "\n",
       "  groundtruth_compartment predicted_compartment  \n",
       "0                 stromal               stromal  \n",
       "1                 stromal               stromal  \n",
       "3                 stromal               stromal  \n",
       "4                 stromal               stromal  \n",
       "7                 stromal               stromal  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hash2kmer_with_predictions = hash2kmer.join(celltype_predictions, on='cell_id')\n",
    "describe(hash2kmer_with_predictions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "automotive-adaptation",
   "metadata": {},
   "outputs": [],
   "source": [
    "hash2kmer_with_predictions['hashval'] = hash2kmer_with_predictions['hashval'].astype(str)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "spare-steam",
   "metadata": {},
   "source": [
    "### Write joined to parquet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "cutting-stick",
   "metadata": {},
   "outputs": [],
   "source": [
    "hash2kmer_with_predictions.to_parquet(\n",
    "    os.path.join(sig_outdir_base, f\"aggregated-hash2kmer-with-predicted-cells__{sketch_id}.parquet\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "latest-coordinator",
   "metadata": {},
   "source": [
    "## Seems like some cell ids aren't there .. why?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "annoying-intensity",
   "metadata": {},
   "outputs": [],
   "source": [
    "hash2kmer_cell_ids = set(hash2kmer.cell_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "electoral-character",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_shared.obs.index = adata_shared.obs.index.str.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "immune-soccer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "61250"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_ids_not_in_hash2kmer = adata_shared.obs.index.difference(hash2kmer_cell_ids)\n",
    "len(cell_ids_not_in_hash2kmer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "id": "primary-following",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Human          27521\n",
       "Mouse lemur    18284\n",
       "Mouse          15372\n",
       "Bat               73\n",
       "Name: species, dtype: int64"
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_shared.obs.loc[cell_ids_not_in_hash2kmer].species.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "therapeutic-range",
   "metadata": {},
   "outputs": [],
   "source": [
    "channel_sizes = adata_shared.obs.groupby('channel').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "id": "fewer-refund",
   "metadata": {},
   "outputs": [],
   "source": [
    "no_id_present_channel_sizes = adata_shared.obs.loc[cell_ids_not_in_hash2kmer].groupby('channel').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "id": "hispanic-alcohol",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "channel\n",
       "10X_P1_6                  7680\n",
       "10X_P1_14                 1413\n",
       "10X_P2_5                   825\n",
       "10X_P3_0                   447\n",
       "10X_P5_13                  936\n",
       "10X_P5_14                 1202\n",
       "10X_P7_8                   294\n",
       "10X_P7_9                   745\n",
       "10X_P8_12                  489\n",
       "10X_P8_13                 1327\n",
       "Antoine_Lung_CD31_10X     5873\n",
       "Antoine_Lung_EPCAM_10X    2647\n",
       "Antoine_Lung_P3_10X       3547\n",
       "Bernard_Lung_10X          2596\n",
       "MACA_18m_F_LUNG_50          14\n",
       "MACA_18m_F_LUNG_51           0\n",
       "MACA_18m_M_LUNG_52           0\n",
       "MACA_18m_M_LUNG_53           0\n",
       "MACA_21m_F_LUNG_54           0\n",
       "MACA_21m_F_LUNG_55           0\n",
       "Martine_Lung_1_10X         856\n",
       "Martine_Lung_2_10X         887\n",
       "P1_2                      1677\n",
       "P1_3                      1932\n",
       "P1_4                      2265\n",
       "P2_1                      2921\n",
       "P2_2                      2764\n",
       "P2_3                      2328\n",
       "P2_4                       882\n",
       "P2_5                      3817\n",
       "P2_6                       922\n",
       "P2_7                      2953\n",
       "P2_8                      2519\n",
       "P3_2                         1\n",
       "P3_3                        29\n",
       "P3_4                         0\n",
       "P3_5                      2511\n",
       "P3_6                         0\n",
       "P3_7                         0\n",
       "Stumpy_Lung_10X           1878\n",
       "bat2-LU                     55\n",
       "bat3-LU                     18\n",
       "dtype: int64"
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_id_present_channel_sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "proper-nigeria",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "channel\n",
       "10X_P1_6                  100.000000\n",
       "10X_P1_14                 100.000000\n",
       "10X_P2_5                  100.000000\n",
       "10X_P3_0                  100.000000\n",
       "10X_P5_13                 100.000000\n",
       "10X_P5_14                 100.000000\n",
       "10X_P7_8                  100.000000\n",
       "10X_P7_9                  100.000000\n",
       "10X_P8_12                 100.000000\n",
       "10X_P8_13                 100.000000\n",
       "Antoine_Lung_CD31_10X     100.000000\n",
       "Antoine_Lung_EPCAM_10X    100.000000\n",
       "Antoine_Lung_P3_10X        81.521489\n",
       "Bernard_Lung_10X          100.000000\n",
       "MACA_18m_F_LUNG_50          5.000000\n",
       "MACA_18m_F_LUNG_51          0.000000\n",
       "MACA_18m_M_LUNG_52          0.000000\n",
       "MACA_18m_M_LUNG_53          0.000000\n",
       "MACA_21m_F_LUNG_54          0.000000\n",
       "MACA_21m_F_LUNG_55          0.000000\n",
       "Martine_Lung_1_10X        100.000000\n",
       "Martine_Lung_2_10X        100.000000\n",
       "P1_2                      100.000000\n",
       "P1_3                      100.000000\n",
       "P1_4                      100.000000\n",
       "P2_1                      100.000000\n",
       "P2_2                      100.000000\n",
       "P2_3                      100.000000\n",
       "P2_4                      100.000000\n",
       "P2_5                      100.000000\n",
       "P2_6                      100.000000\n",
       "P2_7                      100.000000\n",
       "P2_8                      100.000000\n",
       "P3_2                        0.104275\n",
       "P3_3                        1.413255\n",
       "P3_4                        0.000000\n",
       "P3_5                       67.228916\n",
       "P3_6                        0.000000\n",
       "P3_7                        0.000000\n",
       "Stumpy_Lung_10X           100.000000\n",
       "bat2-LU                     2.789047\n",
       "bat3-LU                     0.310131\n",
       "dtype: float64"
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "100 * no_id_present_channel_sizes/channel_sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "infectious-seating",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "species  groundtruth_celltype        alignment_status\n",
       "bat      Alveolar Epithelial Type 2  aligned             13522422\n",
       "                                     unaligned             759922\n",
       "         B cell                      aligned              1055558\n",
       "                                     unaligned              58700\n",
       "         Capillary                   aligned             14688263\n",
       "Name: alignment_status, dtype: int64"
      ]
     },
     "execution_count": 254,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "celltype_hashvals = hash2kmer.groupby(['species', 'groundtruth_celltype']).alignment_status.value_counts()\n",
    "celltype_hashvals.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "accredited-steps",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "species  groundtruth_celltype             alignment_status\n",
       "bat      Alveolar Epithelial Type 2       aligned             13522422\n",
       "                                          unaligned             759922\n",
       "         B cell                           aligned              1055558\n",
       "                                          unaligned              58700\n",
       "         Capillary                        aligned             14688263\n",
       "                                                                ...   \n",
       "mouse    Natural Killer T cell            unaligned              42411\n",
       "         Smooth Muscle and Myofibroblast  aligned               184572\n",
       "                                          unaligned               5618\n",
       "         T cell                           aligned              5653466\n",
       "                                          unaligned             189747\n",
       "Name: alignment_status, Length: 78, dtype: int64"
      ]
     },
     "execution_count": 255,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "celltype_hashvals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "roman-clerk",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>groundtruth_celltype</th>\n",
       "      <th>alignment_status</th>\n",
       "      <th>percentage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bat</td>\n",
       "      <td>Alveolar Epithelial Type 2</td>\n",
       "      <td>aligned</td>\n",
       "      <td>94.679291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bat</td>\n",
       "      <td>Alveolar Epithelial Type 2</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>5.320709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bat</td>\n",
       "      <td>B cell</td>\n",
       "      <td>aligned</td>\n",
       "      <td>94.731920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>bat</td>\n",
       "      <td>B cell</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>5.268080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bat</td>\n",
       "      <td>Capillary</td>\n",
       "      <td>aligned</td>\n",
       "      <td>95.639091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>mouse</td>\n",
       "      <td>Natural Killer T cell</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>3.501037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>mouse</td>\n",
       "      <td>Smooth Muscle and Myofibroblast</td>\n",
       "      <td>aligned</td>\n",
       "      <td>97.046112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>mouse</td>\n",
       "      <td>Smooth Muscle and Myofibroblast</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>2.953888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>mouse</td>\n",
       "      <td>T cell</td>\n",
       "      <td>aligned</td>\n",
       "      <td>96.752694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>mouse</td>\n",
       "      <td>T cell</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>3.247306</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   species             groundtruth_celltype alignment_status  percentage\n",
       "0      bat       Alveolar Epithelial Type 2          aligned   94.679291\n",
       "1      bat       Alveolar Epithelial Type 2        unaligned    5.320709\n",
       "2      bat                           B cell          aligned   94.731920\n",
       "3      bat                           B cell        unaligned    5.268080\n",
       "4      bat                        Capillary          aligned   95.639091\n",
       "..     ...                              ...              ...         ...\n",
       "73   mouse            Natural Killer T cell        unaligned    3.501037\n",
       "74   mouse  Smooth Muscle and Myofibroblast          aligned   97.046112\n",
       "75   mouse  Smooth Muscle and Myofibroblast        unaligned    2.953888\n",
       "76   mouse                           T cell          aligned   96.752694\n",
       "77   mouse                           T cell        unaligned    3.247306\n",
       "\n",
       "[78 rows x 4 columns]"
      ]
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "celltype_hashval_percentage = celltype_hashvals.groupby(level=[0, 1]).apply(lambda x: 100 * x/x.sum())\n",
    "celltype_hashval_percentage.name = 'percentage'\n",
    "celltype_hashval_percentage = celltype_hashval_percentage.reset_index()\n",
    "celltype_hashval_percentage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "id": "conceptual-designation",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_hashval_percentage_unaligned = celltype_hashval_percentage.query('alignment_status == \"unaligned\"')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "id": "ecological-clock",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9                          Fibroblast\n",
       "5                           Capillary\n",
       "17    Smooth Muscle and Myofibroblast\n",
       "11                         Macrophage\n",
       "19                             T cell\n",
       "13                           Monocyte\n",
       "15              Natural Killer T cell\n",
       "3                              B cell\n",
       "1          Alveolar Epithelial Type 2\n",
       "7                           Dendritic\n",
       "Name: groundtruth_celltype, dtype: category\n",
       "Categories (10, object): ['Alveolar Epithelial Type 2', 'B cell', 'Capillary', 'Dendritic', ..., 'Monocyte', 'Natural Killer T cell', 'Smooth Muscle and Myofibroblast', 'T cell']"
      ]
     },
     "execution_count": 259,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "order = celltype_hashval_percentage_unaligned.query('species == \"bat\"').sort_values(\n",
    "    \"percentage\"\n",
    ")[\"groundtruth_celltype\"]\n",
    "order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "violent-merchandise",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.catplot(\n",
    "    data=celltype_hashval_percentage_unaligned,\n",
    "    y=\"groundtruth_celltype\",\n",
    "    x=\"percentage\",\n",
    "    hue=\"species\",\n",
    "    kind=\"bar\",\n",
    "    order=order\n",
    ")\n",
    "g.set(xscale=\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "related-wrestling",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.catplot(\n",
    "    data=celltype_hashval_percentage_unaligned,\n",
    "    y=\"groundtruth_celltype\",\n",
    "    x=\"percentage\",\n",
    "    col=\"species\",\n",
    "    kind=\"bar\",\n",
    "    order=order\n",
    ")\n",
    "g.set(xscale=\"log\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "great-boundary",
   "metadata": {},
   "source": [
    "## Read mouse diagnostic k-mer csvs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "pursuant-coral",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/home/olga/data_sm/immune-evolution/kmer-signatures/1--train-mouse/5--celltype-kmers--merged-celltype-remove-common-kmers--min-kmer-count--5-percent/alphabet-dayhoff__ksize-51__scaled-10/csvs/natural_killer_t_cell.csv',\n",
       " '/home/olga/data_sm/immune-evolution/kmer-signatures/1--train-mouse/5--celltype-kmers--merged-celltype-remove-common-kmers--min-kmer-count--5-percent/alphabet-dayhoff__ksize-51__scaled-10/csvs/alveolar_epithelial_type_2.csv',\n",
       " '/home/olga/data_sm/immune-evolution/kmer-signatures/1--train-mouse/5--celltype-kmers--merged-celltype-remove-common-kmers--min-kmer-count--5-percent/alphabet-dayhoff__ksize-51__scaled-10/csvs/smooth_muscle_and_myofibroblast.csv',\n",
       " '/home/olga/data_sm/immune-evolution/kmer-signatures/1--train-mouse/5--celltype-kmers--merged-celltype-remove-common-kmers--min-kmer-count--5-percent/alphabet-dayhoff__ksize-51__scaled-10/csvs/monocyte.csv',\n",
       " '/home/olga/data_sm/immune-evolution/kmer-signatures/1--train-mouse/5--celltype-kmers--merged-celltype-remove-common-kmers--min-kmer-count--5-percent/alphabet-dayhoff__ksize-51__scaled-10/csvs/b_cell.csv',\n",
       " '/home/olga/data_sm/immune-evolution/kmer-signatures/1--train-mouse/5--celltype-kmers--merged-celltype-remove-common-kmers--min-kmer-count--5-percent/alphabet-dayhoff__ksize-51__scaled-10/csvs/fibroblast.csv',\n",
       " '/home/olga/data_sm/immune-evolution/kmer-signatures/1--train-mouse/5--celltype-kmers--merged-celltype-remove-common-kmers--min-kmer-count--5-percent/alphabet-dayhoff__ksize-51__scaled-10/csvs/t_cell.csv',\n",
       " '/home/olga/data_sm/immune-evolution/kmer-signatures/1--train-mouse/5--celltype-kmers--merged-celltype-remove-common-kmers--min-kmer-count--5-percent/alphabet-dayhoff__ksize-51__scaled-10/csvs/capillary.csv',\n",
       " '/home/olga/data_sm/immune-evolution/kmer-signatures/1--train-mouse/5--celltype-kmers--merged-celltype-remove-common-kmers--min-kmer-count--5-percent/alphabet-dayhoff__ksize-51__scaled-10/csvs/dendritic.csv',\n",
       " '/home/olga/data_sm/immune-evolution/kmer-signatures/1--train-mouse/5--celltype-kmers--merged-celltype-remove-common-kmers--min-kmer-count--5-percent/alphabet-dayhoff__ksize-51__scaled-10/csvs/macrophage.csv']"
      ]
     },
     "execution_count": 261,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_dir = '/home/olga/data_sm/immune-evolution/kmer-signatures/1--train-mouse/'\n",
    "celltype_kmer_subdir = '5--celltype-kmers--merged-celltype-remove-common-kmers--min-kmer-count--5-percent'\n",
    "\n",
    "mouse_celltype_kmer_csvs = glob.glob(os.path.join(mouse_dir, celltype_kmer_subdir, sketch_id, 'csvs', '*csv'))\n",
    "mouse_celltype_kmer_csvs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "demographic-doctor",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "natural_killer_t_cell.csv\n",
      "CPU times: user 1.84 s, sys: 182 ms, total: 2.02 s\n",
      "Wall time: 2.54 s\n",
      "alveolar_epithelial_type_2.csv\n",
      "CPU times: user 1.51 s, sys: 142 ms, total: 1.65 s\n",
      "Wall time: 2.06 s\n",
      "smooth_muscle_and_myofibroblast.csv\n",
      "CPU times: user 514 ms, sys: 60.8 ms, total: 575 ms\n",
      "Wall time: 881 ms\n",
      "monocyte.csv\n",
      "CPU times: user 27.5 s, sys: 2.58 s, total: 30.1 s\n",
      "Wall time: 35.9 s\n",
      "b_cell.csv\n",
      "CPU times: user 8.55 s, sys: 915 ms, total: 9.47 s\n",
      "Wall time: 16.3 s\n",
      "fibroblast.csv\n",
      "CPU times: user 3.84 s, sys: 579 ms, total: 4.42 s\n",
      "Wall time: 5.48 s\n",
      "t_cell.csv\n",
      "CPU times: user 5.14 s, sys: 518 ms, total: 5.66 s\n",
      "Wall time: 6.77 s\n",
      "capillary.csv\n",
      "CPU times: user 11.8 s, sys: 1.35 s, total: 13.1 s\n",
      "Wall time: 16 s\n",
      "dendritic.csv\n",
      "CPU times: user 2.54 s, sys: 253 ms, total: 2.8 s\n",
      "Wall time: 3.4 s\n",
      "macrophage.csv\n",
      "CPU times: user 2.42 s, sys: 226 ms, total: 2.65 s\n",
      "Wall time: 3.91 s\n",
      "(22370943, 4)\n",
      "--- First 5 entries ---\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer_in_sequence</th>\n",
       "      <th>kmer_in_alphabet</th>\n",
       "      <th>hashval</th>\n",
       "      <th>read_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KLSSPRGGMKKKTTHFG</td>\n",
       "      <td>debbbdbbedddbbdfb</td>\n",
       "      <td>308103755003737144</td>\n",
       "      <td>A00111:73:H3YHNDMXX:2:1218:21983:15264\\tCB:Z:A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LSSPRGGMKKKTTHFGE</td>\n",
       "      <td>ebbbdbbedddbbdfbc</td>\n",
       "      <td>177719887282804153</td>\n",
       "      <td>A00111:73:H3YHNDMXX:2:1218:21983:15264\\tCB:Z:A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YTVGKRFKEANNFLRPF</td>\n",
       "      <td>fbebddfdcbccfedbf</td>\n",
       "      <td>1822116523473074621</td>\n",
       "      <td>A00111:73:H3YHNDMXX:1:2311:5367:19179\\tCB:Z:AA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KFGIICMEDLIQDIYTI</td>\n",
       "      <td>dfbeeaecceeccefbe</td>\n",
       "      <td>115722640331910096</td>\n",
       "      <td>A00111:73:H3YHNDMXX:2:1175:17463:20556\\tCB:Z:A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KFGIICMEDLIQEIYTV</td>\n",
       "      <td>dfbeeaecceeccefbe</td>\n",
       "      <td>115722640331910096</td>\n",
       "      <td>A00111:73:H3YHNDMXX:2:2229:8260:32111\\tCB:Z:AA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    kmer_in_sequence   kmer_in_alphabet              hashval  \\\n",
       "0  KLSSPRGGMKKKTTHFG  debbbdbbedddbbdfb   308103755003737144   \n",
       "1  LSSPRGGMKKKTTHFGE  ebbbdbbedddbbdfbc   177719887282804153   \n",
       "2  YTVGKRFKEANNFLRPF  fbebddfdcbccfedbf  1822116523473074621   \n",
       "3  KFGIICMEDLIQDIYTI  dfbeeaecceeccefbe   115722640331910096   \n",
       "4  KFGIICMEDLIQEIYTV  dfbeeaecceeccefbe   115722640331910096   \n",
       "\n",
       "                                           read_name  \n",
       "0  A00111:73:H3YHNDMXX:2:1218:21983:15264\\tCB:Z:A...  \n",
       "1  A00111:73:H3YHNDMXX:2:1218:21983:15264\\tCB:Z:A...  \n",
       "2  A00111:73:H3YHNDMXX:1:2311:5367:19179\\tCB:Z:AA...  \n",
       "3  A00111:73:H3YHNDMXX:2:1175:17463:20556\\tCB:Z:A...  \n",
       "4  A00111:73:H3YHNDMXX:2:2229:8260:32111\\tCB:Z:AA...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 21s, sys: 21.4 s, total: 1min 42s\n",
      "Wall time: 2min 31s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "dfs = []\n",
    "\n",
    "for csv in mouse_celltype_kmer_csvs:\n",
    "    basename = os.path.basename(csv)\n",
    "    print('\\ncelltype file:', )\n",
    "    %time df = pd.read_csv(csv)\n",
    "    df['celltype'] = basename.split('.')[0]\n",
    "    dfs.append(df)\n",
    "mouse_celltype_kmers = pd.concat(dfs)\n",
    "describe(mouse_celltype_kmers)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "personal-witch",
   "metadata": {},
   "source": [
    "# Overlap in hashes for cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "id": "native-showcase",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer_in_sequence</th>\n",
       "      <th>kmer_in_alphabet</th>\n",
       "      <th>hashval</th>\n",
       "      <th>read_name</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>species</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>sketch_id</th>\n",
       "      <th>moltype</th>\n",
       "      <th>ksize</th>\n",
       "      <th>scaled</th>\n",
       "      <th>alignment_status</th>\n",
       "      <th>groundtruth_celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DLNVKPKTIKTLEENLG</td>\n",
       "      <td>cecedbdbedbeccceb</td>\n",
       "      <td>550348301278027135</td>\n",
       "      <td>A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IKTLEENLGITIQDIGV</td>\n",
       "      <td>edbecccebebeccebe</td>\n",
       "      <td>974148631326387115</td>\n",
       "      <td>A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CRKLKLDPFLTPYTKIN</td>\n",
       "      <td>addedecbfebbfbdec</td>\n",
       "      <td>1662986983738689810</td>\n",
       "      <td>A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RKLKLDPFLTPYTKINS</td>\n",
       "      <td>ddedecbfebbfbdecb</td>\n",
       "      <td>799720167227311772</td>\n",
       "      <td>A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>LDPFLTPYTKINSRWIK</td>\n",
       "      <td>ecbfebbfbdecbdfed</td>\n",
       "      <td>537198855570371996</td>\n",
       "      <td>A00111:155:HFCWYDMXX:2:1360:14886:30185\\tCB:Z:...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    kmer_in_sequence   kmer_in_alphabet              hashval  \\\n",
       "0  DLNVKPKTIKTLEENLG  cecedbdbedbeccceb   550348301278027135   \n",
       "1  IKTLEENLGITIQDIGV  edbecccebebeccebe   974148631326387115   \n",
       "3  CRKLKLDPFLTPYTKIN  addedecbfebbfbdec  1662986983738689810   \n",
       "4  RKLKLDPFLTPYTKINS  ddedecbfebbfbdecb   799720167227311772   \n",
       "7  LDPFLTPYTKINSRWIK  ecbfebbfbdecbdfed   537198855570371996   \n",
       "\n",
       "                                           read_name gene_name species  \\\n",
       "0  A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...      None   human   \n",
       "1  A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...      None   human   \n",
       "3  A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...      None   human   \n",
       "4  A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...      None   human   \n",
       "7  A00111:155:HFCWYDMXX:2:1360:14886:30185\\tCB:Z:...      None   human   \n",
       "\n",
       "                  cell_id                              sketch_id  moltype  \\\n",
       "0  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "1  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "3  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "4  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "7  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "\n",
       "  ksize  scaled alignment_status groundtruth_celltype  \n",
       "0    51      10        unaligned           Fibroblast  \n",
       "1    51      10        unaligned           Fibroblast  \n",
       "3    51      10        unaligned           Fibroblast  \n",
       "4    51      10        unaligned           Fibroblast  \n",
       "7    51      10        unaligned           Fibroblast  "
      ]
     },
     "execution_count": 263,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hash2kmer.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "knowing-infrared",
   "metadata": {},
   "source": [
    "## Remove lemur for now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "measured-wallpaper",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer_in_sequence</th>\n",
       "      <th>kmer_in_alphabet</th>\n",
       "      <th>hashval</th>\n",
       "      <th>read_name</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>species</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>sketch_id</th>\n",
       "      <th>moltype</th>\n",
       "      <th>ksize</th>\n",
       "      <th>scaled</th>\n",
       "      <th>alignment_status</th>\n",
       "      <th>groundtruth_celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DLNVKPKTIKTLEENLG</td>\n",
       "      <td>cecedbdbedbeccceb</td>\n",
       "      <td>550348301278027135</td>\n",
       "      <td>A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IKTLEENLGITIQDIGV</td>\n",
       "      <td>edbecccebebeccebe</td>\n",
       "      <td>974148631326387115</td>\n",
       "      <td>A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CRKLKLDPFLTPYTKIN</td>\n",
       "      <td>addedecbfebbfbdec</td>\n",
       "      <td>1662986983738689810</td>\n",
       "      <td>A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RKLKLDPFLTPYTKINS</td>\n",
       "      <td>ddedecbfebbfbdecb</td>\n",
       "      <td>799720167227311772</td>\n",
       "      <td>A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>LDPFLTPYTKINSRWIK</td>\n",
       "      <td>ecbfebbfbdecbdfed</td>\n",
       "      <td>537198855570371996</td>\n",
       "      <td>A00111:155:HFCWYDMXX:2:1360:14886:30185\\tCB:Z:...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_4__GAGCAGAAGTGGACGT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    kmer_in_sequence   kmer_in_alphabet              hashval  \\\n",
       "0  DLNVKPKTIKTLEENLG  cecedbdbedbeccceb   550348301278027135   \n",
       "1  IKTLEENLGITIQDIGV  edbecccebebeccebe   974148631326387115   \n",
       "3  CRKLKLDPFLTPYTKIN  addedecbfebbfbdec  1662986983738689810   \n",
       "4  RKLKLDPFLTPYTKINS  ddedecbfebbfbdecb   799720167227311772   \n",
       "7  LDPFLTPYTKINSRWIK  ecbfebbfbdecbdfed   537198855570371996   \n",
       "\n",
       "                                           read_name gene_name species  \\\n",
       "0  A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...      None   human   \n",
       "1  A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...      None   human   \n",
       "3  A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...      None   human   \n",
       "4  A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...      None   human   \n",
       "7  A00111:155:HFCWYDMXX:2:1360:14886:30185\\tCB:Z:...      None   human   \n",
       "\n",
       "                  cell_id                              sketch_id  moltype  \\\n",
       "0  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "1  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "3  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "4  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "7  P3_4__GAGCAGAAGTGGACGT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "\n",
       "  ksize  scaled alignment_status groundtruth_celltype  \n",
       "0    51      10        unaligned           Fibroblast  \n",
       "1    51      10        unaligned           Fibroblast  \n",
       "3    51      10        unaligned           Fibroblast  \n",
       "4    51      10        unaligned           Fibroblast  \n",
       "7    51      10        unaligned           Fibroblast  "
      ]
     },
     "execution_count": 264,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hash2kmer_no_lemur = hash2kmer.query('species != \"lemur\"')\n",
    "hash2kmer_no_lemur.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "historical-mystery",
   "metadata": {},
   "source": [
    "## get diagnostic k-mers for each species and "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "forty-ladder",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method DataFrame.groupby of         kmer_in_sequence   kmer_in_alphabet              hashval  \\\n",
       "0      DLNVKPKTIKTLEENLG  cecedbdbedbeccceb   550348301278027135   \n",
       "1      IKTLEENLGITIQDIGV  edbecccebebeccebe   974148631326387115   \n",
       "3      CRKLKLDPFLTPYTKIN  addedecbfebbfbdec  1662986983738689810   \n",
       "4      RKLKLDPFLTPYTKINS  ddedecbfebbfbdecb   799720167227311772   \n",
       "7      LDPFLTPYTKINSRWIK  ecbfebbfbdecbdfed   537198855570371996   \n",
       "...                  ...                ...                  ...   \n",
       "23409  FIWNPRLSKVSHRSSSV  fefcbdebdebddbbbe   989408111628569969   \n",
       "23410  VSHRSSSVLSHGLCVDT  ebddbbbeebdbeaecb    62515329136865349   \n",
       "23413  VFLFLSSSLPFLFSLFH  efefebbbebfefbefd  1563197722135732861   \n",
       "23415  SSSLPFLFSHFPMYSAL  bbbebfefbdfbefbbe   605995314189153464   \n",
       "23416  PFLFSHFPMYSALIPLL  bfefbdfbefbbeebee   741982259205681112   \n",
       "\n",
       "                                               read_name gene_name species  \\\n",
       "0      A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...      None   human   \n",
       "1      A00111:155:HFCWYDMXX:1:1359:1190:25066\\tCB:Z:G...      None   human   \n",
       "3      A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...      None   human   \n",
       "4      A00111:155:HFCWYDMXX:2:1129:25961:14544\\tCB:Z:...      None   human   \n",
       "7      A00111:155:HFCWYDMXX:2:1360:14886:30185\\tCB:Z:...      None   human   \n",
       "...                                                  ...       ...     ...   \n",
       "23409  A00111:78:H3YHCDMXX:2:1219:31756:32189\\tCB:Z:C...      None   mouse   \n",
       "23410  A00111:78:H3YHCDMXX:2:1219:31756:32189\\tCB:Z:C...      None   mouse   \n",
       "23413  A00111:78:H3YHCDMXX:2:1117:25988:16971\\tCB:Z:C...      None   mouse   \n",
       "23415  A00111:78:H3YHCDMXX:1:1301:32606:26866\\tCB:Z:C...      None   mouse   \n",
       "23416  A00111:78:H3YHCDMXX:1:1301:32606:26866\\tCB:Z:C...      None   mouse   \n",
       "\n",
       "                                    cell_id  \\\n",
       "0                    P3_4__GAGCAGAAGTGGACGT   \n",
       "1                    P3_4__GAGCAGAAGTGGACGT   \n",
       "3                    P3_4__GAGCAGAAGTGGACGT   \n",
       "4                    P3_4__GAGCAGAAGTGGACGT   \n",
       "7                    P3_4__GAGCAGAAGTGGACGT   \n",
       "...                                     ...   \n",
       "23409  MACA_21M_F_LUNG_55__CACACAATCCTCAACC   \n",
       "23410  MACA_21M_F_LUNG_55__CACACAATCCTCAACC   \n",
       "23413  MACA_21M_F_LUNG_55__CACACAATCCTCAACC   \n",
       "23415  MACA_21M_F_LUNG_55__CACACAATCCTCAACC   \n",
       "23416  MACA_21M_F_LUNG_55__CACACAATCCTCAACC   \n",
       "\n",
       "                                   sketch_id  moltype ksize  scaled  \\\n",
       "0      alphabet-dayhoff__ksize-51__scaled-10  dayhoff    51      10   \n",
       "1      alphabet-dayhoff__ksize-51__scaled-10  dayhoff    51      10   \n",
       "3      alphabet-dayhoff__ksize-51__scaled-10  dayhoff    51      10   \n",
       "4      alphabet-dayhoff__ksize-51__scaled-10  dayhoff    51      10   \n",
       "7      alphabet-dayhoff__ksize-51__scaled-10  dayhoff    51      10   \n",
       "...                                      ...      ...   ...     ...   \n",
       "23409  alphabet-dayhoff__ksize-51__scaled-10  dayhoff    51      10   \n",
       "23410  alphabet-dayhoff__ksize-51__scaled-10  dayhoff    51      10   \n",
       "23413  alphabet-dayhoff__ksize-51__scaled-10  dayhoff    51      10   \n",
       "23415  alphabet-dayhoff__ksize-51__scaled-10  dayhoff    51      10   \n",
       "23416  alphabet-dayhoff__ksize-51__scaled-10  dayhoff    51      10   \n",
       "\n",
       "      alignment_status        groundtruth_celltype  \n",
       "0            unaligned                  Fibroblast  \n",
       "1            unaligned                  Fibroblast  \n",
       "3            unaligned                  Fibroblast  \n",
       "4            unaligned                  Fibroblast  \n",
       "7            unaligned                  Fibroblast  \n",
       "...                ...                         ...  \n",
       "23409          aligned  Alveolar Epithelial Type 2  \n",
       "23410          aligned  Alveolar Epithelial Type 2  \n",
       "23413          aligned  Alveolar Epithelial Type 2  \n",
       "23415          aligned  Alveolar Epithelial Type 2  \n",
       "23416          aligned  Alveolar Epithelial Type 2  \n",
       "\n",
       "[146094013 rows x 13 columns]>"
      ]
     },
     "execution_count": 265,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hash2kmer_no_lemur.groupby"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "id": "wicked-coverage",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'broad_group'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-266-f1b1d0deb4b8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mcelltype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mhash2kmer_no_lemur\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"broad_group\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0mspecies_hashes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'species'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkmer_in_alphabet\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/immune-evolution/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mgroupby\u001b[0;34m(self, by, axis, level, as_index, sort, group_keys, squeeze, observed, dropna)\u001b[0m\n\u001b[1;32m   6518\u001b[0m             \u001b[0msqueeze\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msqueeze\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6519\u001b[0m             \u001b[0mobserved\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mobserved\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6520\u001b[0;31m             \u001b[0mdropna\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdropna\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6521\u001b[0m         )\n\u001b[1;32m   6522\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/immune-evolution/lib/python3.7/site-packages/pandas/core/groupby/groupby.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, obj, keys, axis, level, grouper, exclusions, selection, as_index, sort, group_keys, squeeze, observed, mutated, dropna)\u001b[0m\n\u001b[1;32m    531\u001b[0m                 \u001b[0mobserved\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mobserved\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    532\u001b[0m                 \u001b[0mmutated\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmutated\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 533\u001b[0;31m                 \u001b[0mdropna\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropna\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    534\u001b[0m             )\n\u001b[1;32m    535\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/immune-evolution/lib/python3.7/site-packages/pandas/core/groupby/grouper.py\u001b[0m in \u001b[0;36mget_grouper\u001b[0;34m(obj, key, axis, level, sort, observed, mutated, validate, dropna)\u001b[0m\n\u001b[1;32m    779\u001b[0m                 \u001b[0min_axis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgpr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgpr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    780\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 781\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgpr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    782\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgpr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mGrouper\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mgpr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkey\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    783\u001b[0m             \u001b[0;31m# Add key to exclusions\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'broad_group'"
     ]
    }
   ],
   "source": [
    "for celltype, df in hash2kmer_no_lemur.groupby(\"broad_group\"):\n",
    "    species_hashes = df.groupby('species').kmer_in_alphabet.apply(lambda x: set(x)).to_dict()\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "loving-quarter",
   "metadata": {},
   "source": [
    "### Matplotlib venn diagrams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "pressing-breakdown",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib_venn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "divine-clearance",
   "metadata": {},
   "outputs": [],
   "source": [
    "hashes = list(species_hashes.values())\n",
    "# hashes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bottom-scholar",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(hashes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "pediatric-artwork",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "familiar-evaluation",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.groupby('species').cell_id.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "peripheral-colombia",
   "metadata": {},
   "outputs": [],
   "source": [
    "matplotlib_venn.venn3(list(species_hashes.values()), set_labels=list(species_hashes.keys()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "combined-wrapping",
   "metadata": {},
   "source": [
    "### Bat has the most unique hashes per cell type but has the fewest cells??"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "rotary-alert",
   "metadata": {},
   "outputs": [],
   "source": [
    "hash2kmer_no_lemur.groupby(['species']).apply(lambda x: x.hashval.nunique()/x.cell_id.nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "stretch-diary",
   "metadata": {},
   "source": [
    "## Try UpSet Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ahead-appointment",
   "metadata": {},
   "outputs": [],
   "source": [
    "import upsetplot as up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "experimental-jason",
   "metadata": {},
   "outputs": [],
   "source": [
    "upsetplot_data = up.from_contents(species_hashes)\n",
    "upsetplot_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "minor-heath",
   "metadata": {},
   "outputs": [],
   "source": [
    "upsetplotted = up.plot(upsetplot_data)\n",
    "upsetplotted['intersections'].set_title(celltype)\n",
    "upsetplotted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acceptable-vitamin",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "assigned-mandate",
   "metadata": {},
   "source": [
    "## Do on all celltypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "piano-pearl",
   "metadata": {},
   "outputs": [],
   "source": [
    "for celltype, df in hash2kmer_no_lemur.groupby(\"broad_group\"):\n",
    "    species_hashes = (\n",
    "        df.groupby(\"species\").kmer_in_alphabet.apply(lambda x: set(x)).to_dict()\n",
    "    )\n",
    "    fig, ax = plt.subplots()\n",
    "\n",
    "    matplotlib_venn.venn3(\n",
    "        list(species_hashes.values()), set_labels=list(species_hashes.keys())\n",
    "    )\n",
    "    ax.set(title = celltype)\n",
    "    \n",
    "    upsetplot_data = up.from_contents(species_hashes)\n",
    "    upsetplotted = up.plot(upsetplot_data)\n",
    "    upsetplotted['intersections'].set_title(celltype)\n",
    "#     upsetplotted"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "united-thirty",
   "metadata": {},
   "source": [
    "## Do on all celltypes, including lemur"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bound-reggae",
   "metadata": {},
   "outputs": [],
   "source": [
    "for celltype, df in hash2kmer_with_predictions.groupby(\"predicted_compartment\"):\n",
    "    species_hashes = (\n",
    "        df.groupby(\"species\").kmer_in_alphabet.apply(lambda x: set(x)).to_dict()\n",
    "    )\n",
    "    upsetplot_data = up.from_contents(species_hashes)\n",
    "    upsetplotted = up.plot(upsetplot_data)\n",
    "    upsetplotted['intersections'].set_title(celltype)\n",
    "#     upsetplotted"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "chinese-empty",
   "metadata": {},
   "source": [
    "## Do on all celltypes, separate by aligned/unaligned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "level-corner",
   "metadata": {},
   "outputs": [],
   "source": [
    "hash2kmer_no_lemur['species__alignment_status'] = hash2kmer_no_lemur['species'] + ', ' + hash2kmer_no_lemur['alignment_status']\n",
    "hash2kmer_no_lemur.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "harmful-grade",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for celltype, df in hash2kmer_no_lemur.groupby(\"broad_group\"):\n",
    "#     species_hashes = (\n",
    "#         df.groupby('species__alignment_status').kmer_in_alphabet.apply(lambda x: set(x)).to_dict()\n",
    "#     )\n",
    "\n",
    "#     upsetplot_data = up.from_contents(species_hashes)\n",
    "#     upsetplotted = up.plot(upsetplot_data, sort_by='cardinality')\n",
    "#     ax = upsetplotted['intersections']\n",
    "#     ax.set(title=celltype, yscale='log')\n",
    "# #     upsetplotted"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "thousand-worship",
   "metadata": {},
   "source": [
    "## Get hashes unaligned in bat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "optional-scottish",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4min 40s, sys: 6.86 s, total: 4min 47s\n",
      "Wall time: 4min 47s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer_in_sequence</th>\n",
       "      <th>kmer_in_alphabet</th>\n",
       "      <th>hashval</th>\n",
       "      <th>read_name</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>species</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>sketch_id</th>\n",
       "      <th>moltype</th>\n",
       "      <th>ksize</th>\n",
       "      <th>scaled</th>\n",
       "      <th>alignment_status</th>\n",
       "      <th>broad_group</th>\n",
       "      <th>species__alignment_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EVLARAIRQEKEIKGIQ</td>\n",
       "      <td>ceebdbedccdcedbec</td>\n",
       "      <td>1355044342132838269</td>\n",
       "      <td>A00111:155:HFCWYDMXX:2:1245:18304:18850\\tCB:Z:...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_7__AACCGCGGTTGTACAC</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>T cell</td>\n",
       "      <td>human, unaligned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WDCKLVQPLWKSVWRFL</td>\n",
       "      <td>fcadeecbefdbefdfe</td>\n",
       "      <td>82630740028545526</td>\n",
       "      <td>A00111:155:HFCWYDMXX:1:1379:18249:8140\\tCB:Z:T...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_7__TACTCGCCAGCCTTGG</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>T cell</td>\n",
       "      <td>human, unaligned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PCSWVGRINIVKMAILP</td>\n",
       "      <td>babfebdeceedebeeb</td>\n",
       "      <td>1049665999009068867</td>\n",
       "      <td>A00111:155:HFCWYDMXX:2:2107:7229:26318\\tCB:Z:C...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_7__CTCTACGAGTCACGCC</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>T cell</td>\n",
       "      <td>human, unaligned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISNFSKVSGYKINVQKS</td>\n",
       "      <td>ebcfbdebbfdececdb</td>\n",
       "      <td>859122113661270444</td>\n",
       "      <td>A00111:155:HFCWYDMXX:1:2130:13295:10864\\tCB:Z:...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_7__CGAATGTAGCGCCTCA</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>T cell</td>\n",
       "      <td>human, unaligned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>WWDCKLVQPLWKSVWRF</td>\n",
       "      <td>ffcadeecbefdbefdf</td>\n",
       "      <td>1029003803404016114</td>\n",
       "      <td>A00111:155:HFCWYDMXX:1:1282:19226:10770\\tCB:Z:...</td>\n",
       "      <td>None</td>\n",
       "      <td>human</td>\n",
       "      <td>P3_7__CGATCGGAGATGCCTT</td>\n",
       "      <td>alphabet-dayhoff__ksize-51__scaled-10</td>\n",
       "      <td>dayhoff</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>T cell</td>\n",
       "      <td>human, unaligned</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     kmer_in_sequence   kmer_in_alphabet              hashval  \\\n",
       "6   EVLARAIRQEKEIKGIQ  ceebdbedccdcedbec  1355044342132838269   \n",
       "6   WDCKLVQPLWKSVWRFL  fcadeecbefdbefdfe    82630740028545526   \n",
       "2   PCSWVGRINIVKMAILP  babfebdeceedebeeb  1049665999009068867   \n",
       "1   ISNFSKVSGYKINVQKS  ebcfbdebbfdececdb   859122113661270444   \n",
       "22  WWDCKLVQPLWKSVWRF  ffcadeecbefdbefdf  1029003803404016114   \n",
       "\n",
       "                                            read_name gene_name species  \\\n",
       "6   A00111:155:HFCWYDMXX:2:1245:18304:18850\\tCB:Z:...      None   human   \n",
       "6   A00111:155:HFCWYDMXX:1:1379:18249:8140\\tCB:Z:T...      None   human   \n",
       "2   A00111:155:HFCWYDMXX:2:2107:7229:26318\\tCB:Z:C...      None   human   \n",
       "1   A00111:155:HFCWYDMXX:1:2130:13295:10864\\tCB:Z:...      None   human   \n",
       "22  A00111:155:HFCWYDMXX:1:1282:19226:10770\\tCB:Z:...      None   human   \n",
       "\n",
       "                   cell_id                              sketch_id  moltype  \\\n",
       "6   P3_7__AACCGCGGTTGTACAC  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "6   P3_7__TACTCGCCAGCCTTGG  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "2   P3_7__CTCTACGAGTCACGCC  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "1   P3_7__CGAATGTAGCGCCTCA  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "22  P3_7__CGATCGGAGATGCCTT  alphabet-dayhoff__ksize-51__scaled-10  dayhoff   \n",
       "\n",
       "   ksize  scaled alignment_status broad_group species__alignment_status  \n",
       "6     51      10        unaligned      T cell          human, unaligned  \n",
       "6     51      10        unaligned      T cell          human, unaligned  \n",
       "2     51      10        unaligned      T cell          human, unaligned  \n",
       "1     51      10        unaligned      T cell          human, unaligned  \n",
       "22    51      10        unaligned      T cell          human, unaligned  "
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "df_bat_unaligned = df.groupby(\"hashval\").filter(\n",
    "    lambda x: \"bat, unaligned\" in set(x[\"species__alignment_status\"])\n",
    ")\n",
    "df_bat_unaligned.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "floating-foster",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['bat, aligned', 'bat, unaligned', 'human, aligned', 'human, unaligned', 'mouse, aligned', 'mouse, unaligned'])"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_hashes.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "fifty-terrace",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "species  broad_group                    \n",
       "bat      Alveolar Epithelial Type 2          832\n",
       "         B cell                              165\n",
       "         Capillary                          2214\n",
       "         Dendritic                            98\n",
       "         Fibroblast                         1382\n",
       "         Macrophage                         1196\n",
       "         Monocyte                            969\n",
       "         Natural Killer T cell               351\n",
       "         Smooth Muscle and Myofibroblast     120\n",
       "         T cell                              376\n",
       "human    Alveolar Epithelial Type 2         3147\n",
       "         B cell                              174\n",
       "         Capillary                          2093\n",
       "         Dendritic                           438\n",
       "         Fibroblast                          725\n",
       "         Macrophage                         3344\n",
       "         Monocyte                            752\n",
       "         Natural Killer T cell               224\n",
       "         Smooth Muscle and Myofibroblast     587\n",
       "         T cell                             2655\n",
       "mouse    Alveolar Epithelial Type 2           51\n",
       "         B cell                              880\n",
       "         Capillary                          1365\n",
       "         Dendritic                           164\n",
       "         Fibroblast                          357\n",
       "         Macrophage                          280\n",
       "         Monocyte                           1676\n",
       "         Natural Killer T cell               196\n",
       "         Smooth Muscle and Myofibroblast      26\n",
       "         T cell                              889\n",
       "Name: cell_id, dtype: int64"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hash2kmer_no_lemur.groupby(['species', 'broad_group']).cell_id.nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "compliant-increase",
   "metadata": {},
   "source": [
    "### Not *exactly* the same number of cells as originally, but same order of magnitude\n",
    "\n",
    "Bat:\n",
    "```\n",
    "broad_group\n",
    "Alveolar Epithelial Type 2          832\n",
    "B cell                              165\n",
    "Capillary                          2214\n",
    "Dendritic                            98\n",
    "Fibroblast                         1382\n",
    "Macrophage                         1196\n",
    "Monocyte                            969\n",
    "Natural Killer T cell               351\n",
    "Smooth Muscle and Myofibroblast     120\n",
    "T cell                              376\n",
    "Name: cell_id, dtype: int64\n",
    "```\n",
    "\n",
    "Human: \n",
    "```\n",
    "broad_group\n",
    "Alveolar Epithelial Type 2         3150\n",
    "B cell                              174\n",
    "Capillary                          2093\n",
    "Dendritic                           438\n",
    "Fibroblast                          725\n",
    "Macrophage                         3356\n",
    "Monocyte                            752\n",
    "Natural Killer T cell               224\n",
    "Smooth Muscle and Myofibroblast     587\n",
    "T cell                             2655\n",
    "Name: cell_id, dtype: int64\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hearing-multiple",
   "metadata": {},
   "source": [
    "# Gene orthology"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "separated-activation",
   "metadata": {},
   "source": [
    "## Assign figure folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "id": "departmental-mechanism",
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_folder = (\n",
    "    \"/home/olga/googledrive/ImmuneEvolution/figures/kmer_gene_orthology\"\n",
    ")\n",
    "!mkdir -p $figure_folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "understanding-translator",
   "metadata": {},
   "source": [
    "## Use 1:1 orthologs from Anndata of bat-Lemur-human-mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "premium-shoulder",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10560, 9)\n",
      "--- First 5 entries ---\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bat__gene_name</th>\n",
       "      <th>mouse_lemur__gene_name-bat</th>\n",
       "      <th>mouse__gene_name-bat</th>\n",
       "      <th>mouse_lemur__gene_name_x-hlm</th>\n",
       "      <th>mouse__gene_name_x-hlm</th>\n",
       "      <th>gene_ids-lemur-hlm</th>\n",
       "      <th>n_cells-mouse-hlm</th>\n",
       "      <th>mouse_lemur__gene_name_y-hlm</th>\n",
       "      <th>mouse__gene_name_y-hlm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>human__gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMD11</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>Samd11</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>Samd11</td>\n",
       "      <td>gene2117</td>\n",
       "      <td>5381</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>Samd11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOC2L</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>Noc2l</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>Noc2l</td>\n",
       "      <td>gene2116</td>\n",
       "      <td>48156</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>Noc2l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLHL17</th>\n",
       "      <td>KLHL17</td>\n",
       "      <td>KLHL17</td>\n",
       "      <td>Klhl17</td>\n",
       "      <td>KLHL17</td>\n",
       "      <td>Klhl17</td>\n",
       "      <td>gene2115</td>\n",
       "      <td>11262</td>\n",
       "      <td>KLHL17</td>\n",
       "      <td>Klhl17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ISG15</th>\n",
       "      <td>ISG15</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>Isg15</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>Isg15</td>\n",
       "      <td>gene2111</td>\n",
       "      <td>74269</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>Isg15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGRN</th>\n",
       "      <td>AGRN</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>Agrn</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>Agrn</td>\n",
       "      <td>gene2110</td>\n",
       "      <td>23538</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>Agrn</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 bat__gene_name mouse_lemur__gene_name-bat  \\\n",
       "human__gene_name                                             \n",
       "SAMD11                   SAMD11                     SAMD11   \n",
       "NOC2L                     NOC2L                      NOC2L   \n",
       "KLHL17                   KLHL17                     KLHL17   \n",
       "ISG15                     ISG15                      ISG15   \n",
       "AGRN                       AGRN                       AGRN   \n",
       "\n",
       "                 mouse__gene_name-bat mouse_lemur__gene_name_x-hlm  \\\n",
       "human__gene_name                                                     \n",
       "SAMD11                         Samd11                       SAMD11   \n",
       "NOC2L                           Noc2l                        NOC2L   \n",
       "KLHL17                         Klhl17                       KLHL17   \n",
       "ISG15                           Isg15                        ISG15   \n",
       "AGRN                             Agrn                         AGRN   \n",
       "\n",
       "                 mouse__gene_name_x-hlm gene_ids-lemur-hlm  n_cells-mouse-hlm  \\\n",
       "human__gene_name                                                                \n",
       "SAMD11                           Samd11           gene2117               5381   \n",
       "NOC2L                             Noc2l           gene2116              48156   \n",
       "KLHL17                           Klhl17           gene2115              11262   \n",
       "ISG15                             Isg15           gene2111              74269   \n",
       "AGRN                               Agrn           gene2110              23538   \n",
       "\n",
       "                 mouse_lemur__gene_name_y-hlm mouse__gene_name_y-hlm  \n",
       "human__gene_name                                                      \n",
       "SAMD11                                 SAMD11                 Samd11  \n",
       "NOC2L                                   NOC2L                  Noc2l  \n",
       "KLHL17                                 KLHL17                 Klhl17  \n",
       "ISG15                                   ISG15                  Isg15  \n",
       "AGRN                                     AGRN                   Agrn  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "describe(adata_shared.var)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "annual-shell",
   "metadata": {},
   "source": [
    "## Assign k-mer type based on alignment status and gene name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "id": "painful-instruction",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bat 10560\n",
      "lemur 10560\n",
      "human 10560\n",
      "mouse 10560\n"
     ]
    }
   ],
   "source": [
    "PER_SPECIES_ORTHOLOGOUS_GENES = {\n",
    "    \"bat\": set(adata_shared.var[\"bat__gene_name\"]),\n",
    "    \"lemur\": set(adata_shared.var[\"mouse_lemur__gene_name_x-hlm\"]),\n",
    "    \"human\": set(adata_shared.var.index),\n",
    "    \"mouse\": set(adata_shared.var[\"mouse__gene_name-bat\"]),\n",
    "}\n",
    "for k, v in PER_SPECIES_ORTHOLOGOUS_GENES.items():\n",
    "    print(k, len(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "defensive-alcohol",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "def is_orthologous(query_gene, species, per_species_orthologous_genes=PER_SPECIES_ORTHOLOGOUS_GENES):\n",
    "    orthologous_genes = per_species_orthologous_genes[species]\n",
    "    return query_gene in orthologous_genes\n",
    "\n",
    "\n",
    "def assign_kmer_category(row):\n",
    "    if row.alignment_status == 'unaligned':\n",
    "        return 'Not in reference genome'\n",
    "    elif pd.isnull(row.gene_name):\n",
    "        return 'In ref genome, not in a gene'\n",
    "    elif is_orthologous(row.gene_name, row.species):\n",
    "        return 'In ref genome, in a 1:1 orthologous gene'\n",
    "    else:\n",
    "        return 'In ref genome, not in a 1:1 orthologous gene'\n",
    "\n",
    "hash2kmer['kmer_category'] = hash2kmer.apply(assign_kmer_category, axis=1)\n",
    "hash2kmer.groupby(['species', 'kmer_category']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "extended-puzzle",
   "metadata": {},
   "outputs": [],
   "source": [
    "hash2kmer_with_predictions = hash2kmer.join(celltype_predictions, on='cell_id')\n",
    "describe(hash2kmer_with_predictions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "understanding-testing",
   "metadata": {},
   "outputs": [],
   "source": [
    "kmer_categories = (\n",
    "    \"Not in reference genome\",\n",
    "    \"In ref genome, not in a gene\",\n",
    "    \"In ref genome, not in a 1:1 orthologous gene\",\n",
    "    \"In ref genome, in a 1:1 orthologous gene\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "exposed-click",
   "metadata": {},
   "outputs": [],
   "source": [
    "hash2kmer_with_predictions['kmer_category'] = pd.Categorical(hash2kmer_with_predictions['kmer_category'], categories=kmer_categories, ordered=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "operational-resistance",
   "metadata": {},
   "source": [
    "## Write hash2kmer with predictions and orthology to parquet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "interior-scenario",
   "metadata": {},
   "outputs": [],
   "source": [
    "hash2kmer_with_predictions.to_parquet(\n",
    "    os.path.join(sig_outdir_base, \"aggregated-hash2kmer-with-predicted-cells--with-orthology.parquet\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "manual-nature",
   "metadata": {},
   "outputs": [],
   "source": [
    "1+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dynamic-circulation",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "celltype_col = \"groundtruth_celltype\"\n",
    "\n",
    "diagnostic_kmers_n_per_category = hash2kmer_with_predictions.groupby(\n",
    "    [\"species\", celltype_col, \"kmer_category\"]\n",
    ").hashval.nunique()\n",
    "diagnostic_kmers_n_per_category.name = \"percent_kmers\"\n",
    "diagnostic_kmers_n_per_category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "alpha-engineer",
   "metadata": {},
   "outputs": [],
   "source": [
    "diagnostic_kmers_n_per_category_df = diagnostic_kmers_n_per_category.groupby(level=[0, 1]).apply(lambda x: 100*x/x.sum()).reset_index()\n",
    "diagnostic_kmers_n_per_category_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "vietnamese-irrigation",
   "metadata": {},
   "outputs": [],
   "source": [
    "diagnostic_kmers_n_per_category_df.query(\n",
    "    '(species == \"mouse\") and (kmer_category == \"Not in reference genome\")'\n",
    ").sort_values('percent_kmers')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "prescribed-association",
   "metadata": {},
   "outputs": [],
   "source": [
    "diagnostic_kmers_n_per_category_df.query(\n",
    "    '(species == \"mouse\") and (kmer_category == \"Not in reference genome\")'\n",
    ").sort_values('percent_kmers')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "silent-retention",
   "metadata": {},
   "outputs": [],
   "source": [
    "diagnostic_kmers_n_per_category_df.query(\n",
    "    '(species == \"lemur\") and (kmer_category == \"Not in reference genome\")'\n",
    ").sort_values('percent_kmers')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "german-venue",
   "metadata": {},
   "source": [
    "# Plot percentage of kmers in ref genome, etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "recognized-patio",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_order = 'human', 'lemur', 'mouse', 'bat'\n",
    "diagnostic_kmers_n_per_category_df['species'] = pd.Categorical(diagnostic_kmers_n_per_category_df['species'], categories=species_order, ordered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "asian-petroleum",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.catplot(\n",
    "    data=diagnostic_kmers_n_per_category_df,\n",
    "    col=celltype_col,\n",
    "    order=species_order,\n",
    "    col_wrap=5,\n",
    "    y=\"species\",\n",
    "    x=\"percent_kmers\",\n",
    "    hue=\"kmer_category\",\n",
    "    hue_order=kmer_categories,\n",
    "    palette='mako',\n",
    "    kind='bar',\n",
    "    height=2,\n",
    "    linewidth=1,\n",
    "    edgecolor='white'\n",
    "#     legend=True,\n",
    ")\n",
    "g.set_titles('{col_name}')\n",
    "for ax in g.axes.flat:\n",
    "    title = ax.get_title()\n",
    "    if title == 'Smooth Muscle and Myofibroblast':\n",
    "        ax.set_title('Smooth Muscle\\nand Myofibroblast', fontsize=10, pad=-20)\n",
    "pdf = os.path.join(figure_folder, 'unstacked_barplot__col-celltype__y-species__hue-kmer_category.pdf')\n",
    "g.savefig(pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "painful-cassette",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.catplot(\n",
    "    data=diagnostic_kmers_n_per_category_df,\n",
    "    hue=celltype_col,\n",
    "#     col_wrap=2,\n",
    "    y=\"species\",\n",
    "    order=species_order,\n",
    "    x=\"percent_kmers\",\n",
    "    col=\"kmer_category\",\n",
    "    col_order=kmer_categories,\n",
    "    palette='tab10',\n",
    "    kind='bar',\n",
    "    height=2.5,\n",
    "    sharex=True,\n",
    "        linewidth=.5,\n",
    "    edgecolor='white'\n",
    "#     legend=True,\n",
    ")\n",
    "g.set_titles('{col_name}')\n",
    "for ax in g.axes.flat:\n",
    "    if ax.is_last_row():\n",
    "        title = ax.get_title()\n",
    "        title = title.replace(',', ',\\n')\n",
    "        ax.set_title(title, fontsize=10, pad=-20)\n",
    "#     if ax.is_first_col():\n",
    "#         ax.set(xscale='log')\n",
    "pdf = os.path.join(figure_folder, 'unstacked_barplot__col-kmer_category__y-species__hue-celltype.pdf')\n",
    "g.savefig(pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "catholic-pharmaceutical",
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "thirty-theta",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_order_no_lemur = ['mouse', 'human', 'bat']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "focal-greece",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(ncols=5, nrows=2, figsize=(12, 3), sharex=True, sharey=True)\n",
    "\n",
    "\n",
    "for (celltype, df), ax in zip(\n",
    "    diagnostic_kmers_n_per_category_df.groupby(celltype_col), axes.flat\n",
    "):\n",
    "    \n",
    "#     legend = ax.is_last_col() and ax.is_last_row()\n",
    "    # One liner to create a stacked bar chart.\n",
    "    sns.histplot(\n",
    "        df,\n",
    "        y=\"species\",\n",
    "        hue=\"kmer_category\",\n",
    "        weights=\"percent_kmers\",\n",
    "        multiple=\"stack\",\n",
    "        palette=\"mako\",\n",
    "        linewidth=1,\n",
    "        edgecolor=\"white\",\n",
    "        legend=False,\n",
    "        order=species_order_no_lemur,\n",
    "        ax=ax,\n",
    "    )\n",
    "    ax.set(\n",
    "        ylabel=\"species\",\n",
    "        xlabel='Percentage',\n",
    "        title=celltype.replace(\"and\", \"\\nand\"),\n",
    "        yticks=species_order,\n",
    "    )\n",
    "    \n",
    "# Fix the legend so it's not on top of the bars.\n",
    "# legend = ax.get_legend()\n",
    "# legend.set_bbox_to_anchor((1, 1))\n",
    "sns.despine()\n",
    "fig.tight_layout()\n",
    "pdf = os.path.join(figure_folder, 'stacked_barplot__col-celtype__y-species__hue-kmer_category.pdf')\n",
    "fig.savefig(pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "associate-wilderness",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(ncols=2, nrows=5, figsize=(4.5, 6), sharex=True, sharey=True)\n",
    "\n",
    "\n",
    "for (celltype, df), ax in zip(\n",
    "    diagnostic_kmers_n_per_category_df.groupby(celltype_col), axes.flat\n",
    "):\n",
    "    \n",
    "#     legend = ax.is_last_col() and ax.is_last_row()\n",
    "    # One liner to create a stacked bar chart.\n",
    "    sns.histplot(\n",
    "        df,\n",
    "        y=\"species\",\n",
    "        hue=\"kmer_category\",\n",
    "        weights=\"percent_kmers\",\n",
    "        multiple=\"stack\",\n",
    "        palette=\"mako\",\n",
    "        linewidth=1,\n",
    "        edgecolor=\"white\",\n",
    "        legend=False,\n",
    "        order=species_order_no_lemur,\n",
    "        ax=ax,\n",
    "    )\n",
    "    ax.set(\n",
    "        ylabel=\"species\",\n",
    "        xlabel='Percentage',\n",
    "        title=celltype.replace(\"and\", \"\\nand\"),\n",
    "        yticks=species_order,\n",
    "    )\n",
    "    \n",
    "# Fix the legend so it's not on top of the bars.\n",
    "# legend = ax.get_legend()\n",
    "# legend.set_bbox_to_anchor((1, 1))\n",
    "sns.despine()\n",
    "fig.tight_layout()\n",
    "pdf = os.path.join(figure_folder, 'stacked_barplot__col-celtype__y-species__hue-kmer_category__two_column.pdf')\n",
    "fig.savefig(pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deadly-sudan",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(ncols=3, figsize=(12, 4), sharex=True, sharey=True)\n",
    "\n",
    "\n",
    "for (species, df), ax in zip(\n",
    "    diagnostic_kmers_n_per_category_df.groupby(\"species\"), axes.flat\n",
    "):\n",
    "    legend = ax.is_last_col()\n",
    "    # One liner to create a stacked bar chart.\n",
    "    sns.histplot(\n",
    "        df,\n",
    "        y=celltype_col,\n",
    "        hue=\"kmer_category\",\n",
    "        weights=\"percent_kmers\",\n",
    "        multiple=\"stack\",\n",
    "        palette=\"mako\",\n",
    "        linewidth=1,\n",
    "        edgecolor=\"white\",\n",
    "        legend=legend,\n",
    "        ax=ax,\n",
    "    )\n",
    "    ax.set(\n",
    "        ylabel=\"percentage\",\n",
    "        title=species,\n",
    "#         yticks=[\"bat\", \"human\"],\n",
    "    )\n",
    "# Fix the legend so it's not on top of the bars.\n",
    "legend = ax.get_legend()\n",
    "legend.set_bbox_to_anchor((1, 1))\n",
    "\n",
    "sns.despine()\n",
    "fig.tight_layout()\n",
    "\n",
    "pdf = os.path.join(figure_folder, 'stacked_barplot__col-species__y-celltype__hue-kmer_category.pdf')\n",
    "fig.savefig(pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "blocked-pixel",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "constant-amsterdam",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "interstate-garlic",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:immune-evolution]",
   "language": "python",
   "name": "conda-env-immune-evolution-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
